### DISCUSSION TITLE
How to download the file from the FilesService?

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#14318] [Archive]
How to download the file from the FilesService?

With a code like this we just get the file metadata:

```javascript
module.exports = ({ filter }, { services, exceptions }) => {
  /** @type {import('directus/dist/services')} */
  const { FilesService } = services;

  filter('units.items.create', async (input, meta, context) => {
    const filesService = new FilesService({
      knex: context.database,
      schema: context.schema,
    });
  
    const file = await filesService.readOne(input.file);
  
    // ...
  });
};
```

But I want to get the file itself (like a buffer or so).

**Comment by br41nslug** [comment]
You can use `AssetsService.getAsset(ID, opts)` to download the file contents

**Reply by abdonrd** [reply]
Thanks @br41nslug! I didn't know that!

Now I just tried with this code:

```javascript
module.exports = ({ filter }, { services, exceptions }) => {
  /** @type {import('directus/dist/services')} */
  const { AssetsService } = services;

  filter('units.items.create', async (input, meta, context) => {
    const assetsService = new AssetsService({
      knex: context.database,
      schema: context.schema,
    });
  
    // `input.file` is the file UUID
    const file = await assetsService.getAsset(input.file);
  
    // ...
  });
};
```

But I get an error:

```
10:23:07 🚨 Cannot use 'in' operator to search for 'width' in undefined 
TypeError: Cannot use 'in' operator to search for 'width' in undefined
    at /Users/abdonrd/Desktop/platypus-directus/node_modules/directus/dist/utils/transformations.js:40:61
```

(Note: The file it's not an image)

**Reply by br41nslug** [reply]
Ah sorry that was a mistake in my example, the transformation options are required.
I believe if you feed that an empty object that should work
`const file = await assetsService.getAsset(input.file, {});`

**Reply by abdonrd** [reply]
It works, thanks! 👏

**Comment by Heartnett** [comment]
> Ah sorry that was a mistake in my example, the transformation options are required. I believe if you feed that an empty object that should work `const file = await assetsService.getAsset(input.file, {});`

@br41nslug I applied that code and also tried removing the empty object parameter, but I'm getting this error:
```json
{
  "name": "Error",
  "message": "TypeError: Cannot read properties of undefined (reading 'collections')"
}
```

I'm running this code inside of an extension.

**Reply by br41nslug** [reply]
That error sounds like `schema` is missing when creating the `AssetsService`.

**Reply by Heartnett** [reply]
Thanks for the quick reply. I applied the schema parameter to the service constructor, but I'm getting 2 different errors now:

When calling `getAsset` with an empty `transformation` parameter object:
```js
await assetsService.getAsset(docId, {})
```
```json
{
  "name": "Error",
  "message": "TypeError: Cannot read properties of undefined (reading 'transforms')"
}
```
When calling `getAsset` without the `transformation` parameter
```js
await assetsService.getAsset(docId)
```
```json
{
  "name": "Error",
  "message": "TypeError: Cannot destructure property 'transformationParams' of 'undefined' as it is undefined."
}
```

**Reply by br41nslug** [reply]
That is correct (or well to be expected) mentioned above here https://github.com/directus/directus/discussions/14318#discussioncomment-3106700

Ah missed the first part, has `transforms` become required somehow in a recent version? If so just add an empty array there as workaround.

**Reply by Heartnett** [reply]
According to the source code on the main branch - yes, the `transforms` parameter should be optional, but not in v10.6.3.
I got it working like so:
```js
await assetsService.getAsset(docId, { transformationParams: {} })
```
